src/VisualStudio/Core/Def/Implementation/Experimentation/KeybindingResetDetector.cs

```diff
@@ -217,37 +251,87 @@ private void ShowGoldBar()
                               action: InfoBarClose));
         }
 
-        private ReSharperStatus IsReSharperEnabled()
+        /// <summary>
+        /// Returns true if ReSharper is installed, enabled, and not suspended.  
+        /// </summary>
+        private async ValueTask<ReSharperStatus> IsReSharperRunningAsync(ReSharperStatus lastStatus, CancellationToken cancellationToken)
         {
-            AssertIsForeground();
-
             // Quick exit if resharper is either uninstalled or not enabled
-            if (!_resharperExtensionEnabled)
+            if (!_resharperExtensionInstalledAndEnabled)
             {
                 return ReSharperStatus.NotInstalledOrDisabled;
             }
 
-            if (_oleCommandTarget == null)
+            await EnsureOleCommandTargetAsync().ConfigureAwait(false);
+
+           // poll until either suspend or resume botton is available, or until operation is canceled
+            while (true)
             {
-                _oleCommandTarget = _serviceProvider.GetService<IOleCommandTarget, SUIHostCommandDispatcher>();
-            }
+                cancellationToken.ThrowIfCancellationRequested();
 
-            var cmds = new OLECMD[1];
-            cmds[0].cmdID = SuspendId;
-            cmds[0].cmdf = 0;
+                var suspendFlag = await QueryStatusAsync(SuspendId).ConfigureAwait(false);
```

Comment:
Consider moving the inner two queries to an async function that can quickly do both queries on the main thread. This switching to the main thread to execute a query status, switch to a background thread to execute a few comparisons, and then back to the main thread to query again feels like it'll introduce more contention that it will solve.

Reply:
Technically this is correct, but the whole thing is rate limited to one operation per two seconds so it should not be even measurable under any real-world scenario.